packages/playwright-test/src/index.ts

```diff
@@ -505,12 +506,31 @@ export const test = _baseTest.extend<TestFixtures, WorkerFixtures>({
       testInfo.errors.push({ message: prependToError });
   }, { scope: 'test',  _title: 'context' } as any],
 
-  context: async ({ _contextFactory }, use) => {
-    await use(await _contextFactory());
+  _contextReuseEnabled: [!!process.env.PW_REUSE_CONTEXT, { scope: 'worker' }],
+
+  context: async ({ playwright, browser, _contextReuseEnabled, _contextFactory, video, trace }, use, testInfo) => {
+    const canReuseContext = shouldCaptureVideo(normalizeVideoMode(video), testInfo) || shouldCaptureTrace(normalizeTraceMode(trace), testInfo);
+    if (!_contextReuseEnabled && !canReuseContext) {
+      await use(await _contextFactory());
+      return;
+    }
+
+    const defaultContextOptions = (playwright.chromium as any)._defaultContextOptions as BrowserContextOptions;
+    const context = await (browser as any)._newContextForReuse(defaultContextOptions);
+    await use(context);
   },
 
-  page: async ({ context }, use) => {
-    await use(await context.newPage());
+  page: async ({ context, _contextReuseEnabled, playwright }, use) => {
+    if (!_contextReuseEnabled) {
```

Comment:
Should this also check `cannotReuseContext`?

Reply:
I don't think so.